### PR TITLE
GH-66285: fix forking in `asyncio`

### LINKED ISSUE #66285: asyncio: support multiprocessing (support fork)
BPO | [22087](https://bugs.python.org/issue22087)
--- | :---
Nosy | @gvanrossum, @pitrou, @1st1, @thehesiod, @miss-islington
PRs | <li>python/cpython#7208</li><li>python/cpython#7215</li><li>python/cpython#7218</li><li>python/cpython#7226</li><li>python/cpython#7232</li><li>python/cpython#7233</li>
Files | <li>[test_loop.py](https://bugs.python.org/file36117/test_loop.py "Uploaded as text/plain at 2014-07-26.18:01:38 by dan.oreilly"): Test script demonstrating the issue</li><li>[handle_mp_unix.diff](https://bugs.python.org/file36118/handle_mp_unix.diff "Uploaded as text/plain at 2014-07-26.18:20:15 by dan.oreilly"): Patch that makes _UnixDefaultEventLoopPolicy create a new loop object if get_event_loop is called in a forked mp child process</li><li>[handle-mp_unix2.patch](https://bugs.python.org/file36119/handle-mp_unix2.patch "Uploaded as text/plain at 2014-07-26.20:13:57 by dan.oreilly"): Use os.getpid() instead of multiprocessing. Store pid state in Policy instance rather than the Loop instance.</li><li>[handle_mp_unix_with_test.diff](https://bugs.python.org/file36134/handle_mp_unix_with_test.diff "Uploaded as text/plain at 2014-07-27.16:09:52 by dan.oreilly"): Adds a unit test to previous patch</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2014-07-26.18:01:10.150>
labels = ['type-bug', 'expert-asyncio']
title = 'asyncio: support multiprocessing (support fork)'
updated_at = <Date 2018-05-30.00:56:36.541>
user = 'https://bugs.python.org/danoreilly'
```

bugs.python.org fields:
```python
activity = <Date 2018-05-30.00:56:36.541>
actor = 'yselivanov'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2014-07-26.18:01:10.150>
creator = 'dan.oreilly'
dependencies = []
files = ['36117', '36118', '36119', '36134']
hgrepos = []
issue_num = 22087
keywords = ['patch']
message_count = 23.0
messages = ['224082', '224084', '224085', '224097', '224125', '224140', '224143', '224144', '224145', '226698', '235404', '235411', '288327', '297222', '297226', '297227', '297229', '318077', '318092', '318135', '318140', '318143', '318144']
nosy_count = 7.0
nosy_names = ['gvanrossum', 'pitrou', 'zmedico', 'yselivanov', 'thehesiod', 'dan.oreilly', 'miss-islington']
pr_nums = ['7208', '7215', '7218', '7226', '7232', '7233']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue22087'
versions = ['Python 3.4', 'Python 3.5', 'Python 3.6']
```

</p></details>

### Linked PRs
* gh-99539
* gh-99745
* gh-99756
* gh-99769

### LINKED ISSUE #94454: asyncio signal handler receives signals from child processes
**Bug report**

Today I ran into a very weird behaviour in one of my projects, that uses asyncio and multiprocessing.
When I interrupted the program with a Ctrl-C on the terminal, the signal handler set in asyncio with `loop.add_signal_handler` was triggered multiple times! Further inspection showed, that if I manually send a `SIGINT` to one of the worker child processes, it triggered the main process signal handler.

I could trace the issue to the socket used by asyncio to receive the signals via `signal.set_wakeup_fd`, which seems to receive the signal bytes also from the child processes.

Running `signal.set_wakeup_fd(-1)` at least once in the worker processes worked as a workaround for me but cannot be the final solution.

Also, it seems, that other signal operations also can clear the issue (for example some automatically manager child processes did not trigger the bug, as they seem to ignore `SIGINT`).

In my project I used `concurrent.futures.ProcessPoolExecutor`, but I don't think that matters, as long as the signal handler (and wakeup fd) is set before the child is started.

**Your environment**

- CPython versions tested on:
Python 3.10.4
- Operating system and architecture:
Kali Linux rolling with kernel: Debian 5.16.18-1kali1 (2022-04-01) x86_64

**Bug output**
[asyncio_signal_bug.py.txt](https://github.com/python/cpython/files/9021468/asyncio_signal_bug.py.txt)

```
# python3 asyncio_signal_bug.py
394236 Signal Handler set
394237 Worker 1 started
394240 Worker 2 started
^C394236 received signal: Signals.SIGINT
394236 received signal: Signals.SIGINT
394236 received signal: Signals.SIGINT
^C394236 received signal: Signals.SIGINT
394236 received signal: Signals.SIGINT
394236 received signal: Signals.SIGINT

# python3 asyncio_signal_bug.py --set-in-children
394241 Signal Handler set
394242 Worker 1 started
394245 Worker 2 started
^C394241 received signal: Signals.SIGINT
^C394241 received signal: Signals.SIGINT
^C394241 received signal: Signals.SIGINT
^C394241 received signal: Signals.SIGINT
^C394241 received signal: Signals.SIGINT
^C394241 received signal: Signals.SIGINT
```
Notice, that in the first block, for every ^C, there are three signals received, whereas in the second block (if `signal.set_wakeup_fd(-1)` was called in the workers) there is only one.

### PR DESCRIPTION
**Pull request by kumaraditya303**
Closes https://github.com/python/cpython/issues/66285
Closes https://github.com/python/cpython/issues/94454

* Issue: gh-66285

### CODE REVIEW

**Comment by gvanrossum** [comment]
See my comments on the issue, @kumaraditya303 .

**Comment by kumaraditya303** [comment]
> See my comments on the issue,

@gvanrossum: I have added more tests and also added a subprocess test.

File: Lib/test/test_asyncio/test_unix_events.py [hunk 1]
```diff
@@ -1867,5 +1870,97 @@ async def runner():
...
+
+            # Test parent's loop is still functional
+            self.assertEqual(await asyncio.create_task(func()), 42)
+
+        asyncio.run(main())
+
+        self.assertFalse(parent_handled.is_set())
+        self.assertTrue(child_handled.is_set())
+
+    def test_fork_asyncio_run(self):
```
Comment (gvanrossum) [inline Lib/test/test_asyncio/test_unix_events.py:1933]: So I commented out the fix and then ran the tests, and this test and the next (`test_form_asyncio_subprocess`) still pass. That's on macOS, so maybe it's different on Linux. But still this doesn't give me much confidence.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_unix_events.py:1933]: That's expected, these two tests are more of a regression test and it works even without this fix. The other two tests should fail as those test the bugs mentioned in the issues.

Comment (gvanrossum) [reply Lib/test/test_asyncio/test_unix_events.py:1933]: Yes, indeed the other two tests pass. Thinking more about this I think it's good to go. As soon you have your credentials you can merge this yourself!

**Review by gvanrossum — APPROVED** [review]

**Comment by brettcannon** [comment]
Looks like the failure was random as subsequent builds succeeded.

https://buildbot.python.org/all/#/builders/1046

**Comment by vstinner** [comment]
Yeah, I saw the test_traceback failure, but it only occurred once, so I ignored it.

**Pull request merged**